diff --git a/nova/tests/console/test_websocketproxy.py b/nova/tests/console/test_websocketproxy.py
index 66913c2..04f6aa2 100644
--- a/nova/tests/console/test_websocketproxy.py
+++ b/nova/tests/console/test_websocketproxy.py
@@ -17,8 +17,13 @@
 
 import mock
 from oslo.config import cfg
+import testtools
+
+try:
+    from nova.console import websocketproxy
+except ImportError:
+    websocketproxy = None
 
-from nova.console import websocketproxy
 from nova import exception
 from nova import test
 
@@ -27,6 +32,7 @@ CONF = cfg.CONF
 
 class NovaProxyRequestHandlerBaseTestCase(test.TestCase):
 
+    @testtools.skipIf(websocketproxy is None, "websockify not available")
     def setUp(self):
         super(NovaProxyRequestHandlerBaseTestCase, self).setUp()
 
