diff --git a/nova/network/linux_net.py b/nova/network/linux_net.py
index 635d1ae..b2317f0 100644
--- a/nova/network/linux_net.py
+++ b/nova/network/linux_net.py
@@ -899,15 +899,20 @@ def get_dns_hosts(context, network_ref):
     return '\n'.join(hosts)
 
 
-def _add_dnsmasq_accept_rules(dev):
+def _add_dnsmasq_accept_rules(dev, network_ref):
     """Allow DHCP and DNS traffic through to dnsmasq."""
     table = iptables_manager.ipv4['filter']
-    for port in [67, 53]:
-        for proto in ['udp', 'tcp']:
-            args = {'dev': dev, 'port': port, 'proto': proto}
-            table.add_rule('INPUT',
-                           '-i %(dev)s -p %(proto)s -m %(proto)s '
-                           '--dport %(port)s -j ACCEPT' % args)
+    for proto in ['udp', 'tcp']:
+        args = {'dev': dev, 'proto': proto}
+        table.add_rule('INPUT',
+                       '-i %(dev)s -p %(proto)s -m %(proto)s '
+                       '--dport 67 -j ACCEPT' % args)
+    for proto in ['udp', 'tcp']:
+        args = {'dev': dev, 'proto': proto,
+                'cidr': network_ref['cidr']}
+        table.add_rule('INPUT',
+                       '-s %(cidr)s -i %(dev)s -p %(proto)s -m %(proto)s '
+                       '--dport 53 -j ACCEPT' % args)
     iptables_manager.apply()
 
 
@@ -1045,7 +1050,7 @@ def restart_dhcp(context, dev, network_ref):
         if conffile.split('/')[-1] in out:
             try:
                 _execute('kill', '-HUP', pid, run_as_root=True)
-                _add_dnsmasq_accept_rules(dev)
+                _add_dnsmasq_accept_rules(dev, network_ref)
                 return
             except Exception as exc:  # pylint: disable=W0703
                 LOG.error(_('Hupping dnsmasq threw %s'), exc)
@@ -1096,7 +1101,7 @@ def restart_dhcp(context, dev, network_ref):
 
     _execute(*cmd, run_as_root=True)
 
-    _add_dnsmasq_accept_rules(dev)
+    _add_dnsmasq_accept_rules(dev, network_ref)
 
 
 @utils.synchronized('radvd_start')

