diff --git a/nova/context.py b/nova/context.py
index c6aca91..d015b30 100644
--- a/nova/context.py
+++ b/nova/context.py
@@ -61,7 +61,7 @@ class RequestContext(object):
             because they possibly came in from older rpc messages.
         """
         if kwargs:
-            LOG.warn(_('Arguments dropped when creating context: %s') %
+            LOG.info(_('Arguments dropped when creating context: %s') %
                     str(kwargs))
 
         self.user_id = user_id
diff --git a/nova/tests/test_context.py b/nova/tests/test_context.py
index 5f2d72a..a349ee7 100644
--- a/nova/tests/test_context.py
+++ b/nova/tests/test_context.py
@@ -67,7 +67,7 @@ class ContextTestCase(test.NoDBTestCase):
         def fake_warn(log_msg):
             info['log_msg'] = log_msg
 
-        self.stubs.Set(context.LOG, 'warn', fake_warn)
+        self.stubs.Set(context.LOG, 'info', fake_warn)
 
         c = context.RequestContext('user', 'project',
                 extra_arg1='meow', extra_arg2='wuff')

