diff --git a/nova/network/linux_net.py b/nova/network/linux_net.py
index 635d1ae..13b1ddd 100644
--- a/nova/network/linux_net.py
+++ b/nova/network/linux_net.py
@@ -669,7 +669,7 @@ def init_host(ip_range):
     # NOTE(devcamcar): Cloud public SNAT entries and the default
     # SNAT rule for outbound traffic.
 
-    add_snat_rule(ip_range)
+    # add_snat_rule(ip_range)
 
     rules = []
     for snat_range in CONF.force_snat_range:
diff --git a/nova/tests/network/test_manager.py b/nova/tests/network/test_manager.py
index 295f6bd..681da79 100644
--- a/nova/tests/network/test_manager.py
+++ b/nova/tests/network/test_manager.py
@@ -1823,12 +1823,7 @@ class CommonNetworkTestCase(test.TestCase):
                                        linux_net.iptables_manager.ipv4['nat'],
                                        table_name='nat')
 
-        expected_lines = ['[0:0] -A %s-snat -s %s -d 0.0.0.0/0 '
-                          '-j SNAT --to-source %s -o %s'
-                          % (binary_name, networks[0]['cidr'],
-                                          CONF.routing_source_ip,
-                                          CONF.public_interface),
-                          '[0:0] -A %s-POSTROUTING -s %s -d %s/32 -j ACCEPT'
+        expected_lines = ['[0:0] -A %s-POSTROUTING -s %s -d %s/32 -j ACCEPT'
                           % (binary_name, networks[0]['cidr'],
                                           CONF.metadata_host),
                           '[0:0] -A %s-POSTROUTING -s %s -d %s -j ACCEPT'
@@ -1838,11 +1833,6 @@ class CommonNetworkTestCase(test.TestCase):
                           '--ctstate DNAT -j ACCEPT' % (binary_name,
                                                         networks[0]['cidr'],
                                                         networks[0]['cidr']),
-                          '[0:0] -A %s-snat -s %s -d 0.0.0.0/0 '
-                          '-j SNAT --to-source %s -o %s'
-                          % (binary_name, networks[1]['cidr'],
-                                          CONF.routing_source_ip,
-                                          CONF.public_interface),
                           '[0:0] -A %s-POSTROUTING -s %s -d %s/32 -j ACCEPT'
                           % (binary_name, networks[1]['cidr'],
                                           CONF.metadata_host),
@@ -1892,12 +1882,7 @@ class CommonNetworkTestCase(test.TestCase):
                                        table_name='nat')
 
         # Add the new expected rules to the old ones
-        expected_lines += ['[0:0] -A %s-snat -s %s -d 0.0.0.0/0 '
-                           '-j SNAT --to-source %s -o %s'
-                           % (binary_name, new_network['cidr'],
-                                           CONF.routing_source_ip,
-                                           CONF.public_interface),
-                           '[0:0] -A %s-POSTROUTING -s %s -d %s/32 -j ACCEPT'
+        expected_lines += ['[0:0] -A %s-POSTROUTING -s %s -d %s/32 -j ACCEPT'
                            % (binary_name, new_network['cidr'],
                                            CONF.metadata_host),
                            '[0:0] -A %s-POSTROUTING -s %s -d %s -j ACCEPT'

